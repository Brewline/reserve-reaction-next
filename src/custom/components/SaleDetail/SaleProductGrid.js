import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withComponents } from "@reactioncommerce/components-context";
import { CustomPropTypes, preventAccidentalDoubleClick } from "@reactioncommerce/components/utils";

const GridContainer = styled.div`
  box-sizing: border-box;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
`;

const GridItem = styled.div`
  box-sizing: border-box;
  flex-basis: 100%;
  flex-grow: 0;
  margin: 0;
  max-width: 100%;
  padding: 12px;
`;

class SaleProductGrid extends Component {
  static propTypes = {
    /**
     * Labels to use for the various badges. Refer to `BadgeOverlay`'s prop documentation.
     */
    badgeLabels: PropTypes.shape({
      BACKORDER: PropTypes.string,
      BESTSELLER: PropTypes.string,
      LOW_QUANTITY: PropTypes.string,
      SOLD_OUT: PropTypes.string,
      SALE: PropTypes.string
    }),
    /**
     * You can provide a `className` prop that will be applied to the outermost DOM element
     * rendered by this component. We do not recommend using this for styling purposes, but
     * it can be useful as a selector in some situations.
     */
    className: PropTypes.string,
    /**
     * If you've set up a components context using
     * [@reactioncommerce/components-context](https://github.com/reactioncommerce/components-context)
     * (recommended), then this prop will come from there automatically. If you have not
     * set up a components context or you want to override one of the components in a
     * single spot, you can pass in the components prop directly.
     */
    components: PropTypes.shape({
      SaleProductGridItem: CustomPropTypes.component.isRequired
    }).isRequired,
    /**
     * Currency code to display the price for. Product must include a pricing object with the code in `product.pricing`
     */
    currencyCode: PropTypes.string,
    /**
     * Item click handler
     */
    onItemClick: PropTypes.func,
    /**
     * Image to display when product doesn't have a primary image
     */
    placeholderImageURL: PropTypes.string,
    /**
     * Products to display in the grid. Refer to `SaleProductGridItem`'s documentation
     */
    products: PropTypes.arrayOf(PropTypes.object)
  };

  static defaultProps = {
    badgeLabels: null,
    currencyCode: "USD",
    initialSize: {
      width: 325
    },
    onItemClick() {},
    placeholderImageURL: "/resources/placeholder.gif",
    products: []
  };

  handleOnClick = preventAccidentalDoubleClick((event, product) => {
    this.props.onItemClick(event, product);
  });

  render() {
    const {
      badgeLabels,
      className,
      components: { SaleProductGridItem },
      currencyCode,
      onItemClick,
      placeholderImageURL,
      products
    } = this.props;

    const gridItemProps = {
      currencyCode,
      placeholderImageURL,
      onClick: onItemClick
    };

    if (badgeLabels) {
      gridItemProps.badgeLabels = badgeLabels;
    }

    return (
      <GridContainer className={className}>
        {products.map((product, index) => (
          <GridItem key={`grid-item-${index}`} {...this.props}>
            <SaleProductGridItem product={product} {...gridItemProps} />
          </GridItem>
        ))}
      </GridContainer>
    );
  }
}

export default withComponents(SaleProductGrid);
