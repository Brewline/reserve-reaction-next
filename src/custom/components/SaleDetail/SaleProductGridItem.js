// forked from: https://raw.githubusercontent.com/reactioncommerce/reaction-component-library/c0610e8c1048d6305eacb5262cdd13985d892a89/package/src/components/CatalogGridItem/v1/CatalogGridItem.js

import React, { Component } from "react";
import PropTypes from "prop-types";
import { inject, observer } from "mobx-react";
import styled from "styled-components";
import { withComponents } from "@reactioncommerce/components-context";
import {
  addTypographyStyles,
  applyTheme,
  CustomPropTypes,
  preventAccidentalDoubleClick
} from "@reactioncommerce/components/utils";
import { priceByCurrencyCode } from "@reactioncommerce/components/CatalogGridItem/v1/utils";
import withWidth, { isWidthUp } from "@material-ui/core/withWidth";
import withCart from "containers/cart/withCart";
import ProductDetailAddToCart from "components/ProductDetailAddToCart";
import VariantList from "components/VariantList";
import TRACKING from "lib/tracking/constants";
import track from "lib/tracking/track";
import trackCartItems from "lib/tracking/trackCartItems";
import trackProduct from "lib/tracking/trackProduct";
import variantById from "lib/utils/variantById";

const { CART_VIEWED, PRODUCT_ADDED, PRODUCT_VIEWED } = TRACKING;

const ProductMediaWrapper = styled.div`
  background-color: ${applyTheme("CatalogGridItem.mediaBackgroundColor")};
  position: relative;
`;

const ProductInfo = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
  padding-bottom: 0;
  padding-left: 0;
  padding-right: 0;
  padding-top: ${applyTheme("CatalogGridItem.verticalSpacingBetweenImageAndInfo")};
`;

const ProductTitle = styled.aside`
  ${addTypographyStyles("CatalogGridItemProductTitle", "headingTextBold")}
  line-height: 1.125;
`;

const ProductVendor = styled.span`
  ${addTypographyStyles("CatalogGridItemProductVendor", "labelText")}
`;

const PriceContainer = styled.div`
  text-align: right;
`;

@withCart
@withWidth({ initialWidth: "md" })
@inject("routingStore", "uiStore")
@track()
@observer
class SaleProductGridItem extends Component {
  static propTypes = {
    /**
     * Function to add items to a cart.
     * Implementation may be provided by addItemsToCart function from the @withCart decorator
     *
     * @example addItemsToCart(CartItemInput)
     * @type Function
     */
    addItemsToCart: PropTypes.func,
    /**
     * Labels to use for the various badges. Refer to `BadgeOverlay`'s prop documentation.
     */
    badgeLabels: PropTypes.shape({
      BACKORDER: PropTypes.string,
      BESTSELLER: PropTypes.string,
      LOW_QUANTITY: PropTypes.string,
      SOLD_OUT: PropTypes.string,
      SALE: PropTypes.string
    }),
    /**
     * You can provide a `className` prop that will be applied to the outermost DOM element
     * rendered by this component. We do not recommend using this for styling purposes, but
     * it can be useful as a selector in some situations.
     */
    className: PropTypes.string,
    /**
     * If you've set up a components context using
     * [@reactioncommerce/components-context](https://github.com/reactioncommerce/components-context)
     * (recommended), then this prop will come from there automatically. If you have not
     * set up a components context or you want to override one of the components in a
     * single spot, you can pass in the components prop directly.
     */
    components: PropTypes.shape({
      BadgeOverlay: CustomPropTypes.component.isRequired,
      Link: CustomPropTypes.component.isRequired,
      Price: CustomPropTypes.component.isRequired,
      ProgressiveImage: CustomPropTypes.component.isRequired
    }),
    /**
     * Currency code to display the price for. Product must include a pricing object with the code in `product.pricing`
     */
    currencyCode: PropTypes.string.isRequired,
    /**
     * Item click handler
     */
    onClick: PropTypes.func,
    /**
     * Image to display when product doesn't have a primary image
     */
    placeholderImageURL: PropTypes.string,
    /**
     * Product to display
     */
    product: PropTypes.shape({
      title: PropTypes.string,
      slug: PropTypes.string,
      vendor: PropTypes.string,
      primaryImage: PropTypes.shape({
        URLs: PropTypes.shape({
          large: PropTypes.string,
          medium: PropTypes.string,
          small: PropTypes.string,
          thumbnail: PropTypes.string
        })
      }),
      pricing: PropTypes.arrayOf(PropTypes.shape({
        currency: PropTypes.shape({
          code: PropTypes.string
        }),
        displayPrice: PropTypes.string
      })),
      isSoldOut: PropTypes.bool,
      isBackorder: PropTypes.bool,
      isOnSale: PropTypes.bool,
      isLowQuantity: PropTypes.bool,
      isBestseller: PropTypes.bool
    }),
    uiStore: PropTypes.object.isRequired
  };

  static defaultProps = {
    badgeLabels: null,
    onClick() {},
    placeholderImageURL: ""
  };

  state = {
    fit: "cover"
  };

  componentDidMount() {
    const { product } = this.props;

    this._mounted = true;

    this.setImageFit();

    // Select first variant by default
    this.selectVariant(product.variants[0], null, false);
  }

  componentDidUpdate() {
    this.setImageFit();
  }

  componentWillUnmount() {
    this._mounted = false;
  }

  setImageFit = () => {
    // Use cover fit if image is landscape, contain if portrait
    if (typeof Image !== "undefined") {
      const { large } = this.primaryImage.URLs;
      const largeImage = new Image();
      largeImage.src = large;
      largeImage.onload = () => {
        if (this._mounted === false) {
          return;
        }

        let fit = "";
        const { width, height } = largeImage;
        if (height > width) {
          // Image is portrait
          fit = "contain";
        } else {
          // Image is landscape
          fit = "cover";
        }

        if (fit !== this.state.fit) {
          this.setState({ fit });
        }
      };
    }
  };

  get productDetailHref() {
    const { product: { slug } } = this.props;
    const url = `/product/${slug}`;
    return url;
  }

  get primaryImage() {
    const { product: { primaryImage }, placeholderImageURL } = this.props;
    if (!primaryImage) {
      return {
        URLs: {
          thumbnail: placeholderImageURL,
          small: placeholderImageURL,
          medium: placeholderImageURL,
          large: placeholderImageURL
        }
      };
    }
    return primaryImage;
  }

  @trackProduct()
  trackAction() {}

  @trackCartItems()
  trackCartItems() {}

  selectVariant(variant, optionId, trackAction = true) {
    const { uiStore } = this.props;

    // Select the variant, and if it has options, the first option
    const variantId = variant._id;
    let selectOptionId = optionId;
    if (!selectOptionId && variant.options && variant.options.length) {
      selectOptionId = variant.options[0]._id;
    }

    if (trackAction) {
      this.trackAction({ variant, optionId, action: PRODUCT_VIEWED });
    }

    uiStore.setPDPSelectedVariantId(variantId, selectOptionId);
  }

  /**
   * @name handleSelectVariant
   * @summary Called when a variant is selected in the variant list
   * @private
   * @ignore
   * @param {Object} variant The variant object that was selected
   * @returns {undefined} No return
   */
  handleSelectVariant = (variant) => {
    this.selectVariant(variant);
  };

  /**
   * @name handleSelectOption
   * @summary Called when an option is selected in the option list
   * @private
   * @ignore
   * @param {Object} option The option object that was selected
   * @returns {undefined} No return
   */
  handleSelectOption = (option) => {
    const { product, uiStore } = this.props;

    // If we are clicking an option, it must be for the current selected variant
    const variant = product.variants.find((vnt) => vnt._id === uiStore.pdpSelectedVariantId);

    this.selectVariant(variant, option._id);
  };

  handleOnClick = preventAccidentalDoubleClick((event) => {
    this.props.onClick(event, this.props.product);
  });

  /**
   * @name handleAddToCartClick
   * @summary Called when the add to cart button is clicked
   * @private
   * @ignore
   * @param {Number} quantity - A positive integer from 0 to infinity, representing the quantity to add to cart
   * @returns {undefined} No return
   */
  handleAddToCartClick = async (quantity) => {
    const {
      addItemsToCart,
      currencyCode,
      product,
      uiStore: { openCartWithTimeout, pdpSelectedOptionId, pdpSelectedVariantId },
      width
    } = this.props;

    // Get selected variant or variant option
    const selectedVariant = variantById(product.variants, pdpSelectedVariantId);
    const selectedOption = variantById(selectedVariant.options, pdpSelectedOptionId);
    const selectedVariantOrOption = selectedOption || selectedVariant;

    debugger;
    if (selectedVariantOrOption) {
      // Get the price for the currently selected variant or variant option
      const price = priceByCurrencyCode(currencyCode, selectedVariantOrOption.pricing);

      // Call addItemsToCart with an object matching the GraphQL `CartItemInput` schema
      const { data } = await addItemsToCart([
        {
          price: {
            amount: price.price,
            currencyCode
          },
          productConfiguration: {
            productId: product.productId, // Pass the productId, not to be confused with _id
            productVariantId: selectedVariantOrOption.variantId // Pass the variantId, not to be confused with _id
          },
          quantity
        }
      ]);

      // If no errors occurred, track action
      if (data) {
        // The response data will be in either `createCart` or `addCartItems` prop
        // depending on the type of user, either authenticated or anonymous.
        const { cart } = data.createCart || data.addCartItems;
        const { edges: items } = cart.items;

        this.trackAction({
          variant: {
            ...selectedVariant,
            cart_id: cart._id, // eslint-disable-line camelcase
            quantity
          },
          optionId: selectedOption ? selectedOption._id : null,
          action: PRODUCT_ADDED
        });

        // The mini cart popper will open automatically after adding an item to the cart,
        // therefore, a CART_VIEWED event is published.
        // debugger // eslint-disable-line
        this.trackCartItems({ cartItems: items, cartId: cart._id, action: CART_VIEWED }); // eslint-disable-line camelcase
      }
    }
    if (isWidthUp("md", width)) {
      // Open the cart, and close after a 3 second delay
      openCartWithTimeout(3000);
    }
  };

  renderProductMedia() {
    const { components: { ProgressiveImage }, product: { description } } = this.props;
    const { fit } = this.state;

    return (
      <ProductMediaWrapper>
        <ProgressiveImage
          fit={fit}
          altText={description}
          presrc={this.primaryImage.URLs.thumbnail}
          srcs={this.primaryImage.URLs}
        />
      </ProductMediaWrapper>
    );
  }

  renderProductInfo() {
    const {
      components: { Price },
      currencyCode,
      product
    } = this.props;
    const { pageTitle, pricing, title } = product;
    const productPrice = priceByCurrencyCode(currencyCode, pricing) || {};

    return (
      <div>
        <ProductInfo>
          <ProductTitle>{title}</ProductTitle>
          <PriceContainer>
            <Price displayPrice={productPrice.displayPrice} />
          </PriceContainer>
        </ProductInfo>
        <div>
          <ProductVendor>{pageTitle}</ProductVendor>
        </div>
      </div>
    );
  }

  renderAddToCart() {
    const {
      currencyCode,
      product,
      uiStore: { pdpSelectedOptionId, pdpSelectedVariantId }
    } = this.props;
    const { variants } = product;

    return (
      <div>
        <VariantList
          onSelectOption={this.handleSelectOption}
          onSelectVariant={this.handleSelectVariant}
          product={product}
          selectedOptionId={pdpSelectedOptionId}
          selectedVariantId={pdpSelectedVariantId}
          currencyCode={currencyCode}
          variants={variants}
        />
        <ProductDetailAddToCart onClick={this.handleAddToCartClick} />
      </div>
    );
  }

  render() {
    const {
      className,
      badgeLabels,
      components: { BadgeOverlay /* , Link */ },
      product
    } = this.props;

    const badgeProps = { product };

    if (badgeLabels) {
      badgeProps.badgeLabels = badgeLabels;
    }

    return (
      <div className={className}>
        {/* Sale Product Detail Page is not yet supported
          <Link
            href={this.productDetailHref}
            onClick={this.handleOnClick}
          >
        */}
        <BadgeOverlay {...badgeProps}>
          {this.renderProductMedia()}
          {this.renderProductInfo()}
        </BadgeOverlay>
        {/*
          </Link>
        */}
        {this.renderAddToCart()}
      </div>
    );
  }
}

export default withComponents(SaleProductGridItem);
