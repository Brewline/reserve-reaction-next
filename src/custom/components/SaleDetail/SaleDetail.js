import moment from "moment";
import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import withWidth, { isWidthDown } from "@material-ui/core/withWidth";
import { inject, observer } from "mobx-react";
import InlineAlert from "@reactioncommerce/components/InlineAlert/v1";
import track from "lib/tracking/track";
// import Breadcrumbs from "components/Breadcrumbs";
// import SaleDetailAddToCart from "components/SaleDetailAddToCart";
import ProductDetailTitle from "components/ProductDetailTitle";
// import VariantList from "components/VariantList";
import ProductDetailVendor from "components/ProductDetailVendor";
import ProductDetailDescription from "components/ProductDetailDescription";
// import SaleDetailPrice from "components/SaleDetailPrice";
// import MediaGallery from "components/MediaGallery";
import { Router } from "routes";
// import priceByCurrencyCode from "lib/utils/priceByCurrencyCode";
// import variantById from "lib/utils/variantById";
// import trackSale from "lib/tracking/trackSale";
import TRACKING from "lib/tracking/constants";
import trackCartItems from "lib/tracking/trackCartItems";
import SaleProductGrid from "./SaleProductGrid";
import SaleProductGridItem from "./SaleProductGridItem";

const { PRODUCT_VIEWED } = TRACKING;

const styles = (theme) => ({
  section: {
    marginBottom: theme.spacing.unit * 2
  },
  breadcrumbGrid: {
    marginBottom: theme.spacing.unit * 2,
    marginTop: theme.spacing.unit * 2
  },
  info: {
    marginBottom: theme.spacing.unit
  },
  notFoundMessage: {
    color: theme.palette.reaction.black65
  }
});

/**
 * Sale detail component
 * @name SaleDetail
 * @param {Object} props Component props
 * @returns {React.Component} React component node that represents a sale detail view
 */
@withWidth({ initialWidth: "md" })
@withStyles(styles, { withTheme: true, name: "SkSaleDetail" })
@inject("routingStore", "uiStore")
@track()
@observer
class SaleDetail extends Component {
  static propTypes = {
    /**
     * Function to add items to a cart.
     * Implementation may be provided by addItemsToCart function from the @withCart decorator
     *
     * @example addItemsToCart(CartItemInput)
     * @type Function
     */
    addItemsToCart: PropTypes.func,
    classes: PropTypes.object,
    currencyCode: PropTypes.string.isRequired,
    initialSize: PropTypes.object,
    routingStore: PropTypes.object.isRequired,
    sale: PropTypes.object,
    shop: PropTypes.object.isRequired,
    theme: PropTypes.object,
    uiStore: PropTypes.object.isRequired,
    width: PropTypes.string.isRequired
  };

  selectVariant(variant, optionId) {
    const { sale, uiStore } = this.props;

    // Select the variant, and if it has options, the first option
    const variantId = variant._id;
    let selectOptionId = optionId;
    if (!selectOptionId && variant.options && variant.options.length) {
      selectOptionId = variant.options[0]._id;
    }

    this.trackAction({ variant, optionId, action: PRODUCT_VIEWED });

    uiStore.setPDPSelectedVariantId(variantId, selectOptionId);

    Router.pushRoute("sale", {
      slugOrId: sale.slug,
      variantId: selectOptionId || variantId
    }, { replace: true });
  }

  // @trackSale()
  trackAction() {}

  @trackCartItems()
  trackCartItems() {}

  /**
   * @name handleSelectVariant
   * @summary Called when a variant is selected in the variant list
   * @private
   * @ignore
   * @param {Object} variant The variant object that was selected
   * @returns {undefined} No return
   */
  handleSelectVariant = (variant) => {
    this.selectVariant(variant);
  };

  // /**
  //  * @name handleAddToCartClick
  //  * @summary Called when the add to cart button is clicked
  //  * @private
  //  * @ignore
  //  * @param {Number} quantity - A positive integer from 0 to infinity, representing the quantity to add to cart
  //  * @returns {undefined} No return
  //  */
  // handleAddToCartClick = async (quantity) => {
  //   const {
  //     addItemsToCart,
  //     currencyCode,
  //     sale,
  //     uiStore: { openCartWithTimeout, pdpSelectedOptionId, pdpSelectedVariantId },
  //     width
  //   } = this.props;

  //   // Get selected variant or variant option
  //   const selectedVariant = variantById(sale.variants, pdpSelectedVariantId);
  //   const selectedOption = variantById(selectedVariant.options, pdpSelectedOptionId);
  //   const selectedVariantOrOption = selectedOption || selectedVariant;

  //   if (selectedVariantOrOption) {
  //     // Get the price for the currently selected variant or variant option
  //     const price = priceByCurrencyCode(currencyCode, selectedVariantOrOption.pricing);

  //     // Call addItemsToCart with an object matching the GraphQL `CartItemInput` schema
  //     const { data } = await addItemsToCart([
  //       {
  //         price: {
  //           amount: price.price,
  //           currencyCode
  //         },
  //         saleConfiguration: {
  //           saleId: sale.saleId, // Pass the saleId, not to be confused with _id
  //           saleVariantId: selectedVariantOrOption.variantId // Pass the variantId, not to be confused with _id
  //         },
  //         quantity
  //       }
  //     ]);

  //     // If no errors occurred, track action
  //     if (data) {
  //       // The response data will be in either `createCart` or `addCartItems` prop
  //       // depending on the type of user, either authenticated or anonymous.
  //       const { cart } = data.createCart || data.addCartItems;
  //       const { edges: items } = cart.items;

  //       this.trackAction({
  //         variant: {
  //           ...selectedVariant,
  //           cart_id: cart._id, // eslint-disable-line camelcase
  //           quantity
  //         },
  //         optionId: selectedOption ? selectedOption._id : null,
  //         action: PRODUCT_ADDED
  //       });

  //       // The mini cart popper will open automatically after adding an item to the cart,
  //       // therefore, a CART_VIEWED event is published.
  //       // debugger // eslint-disable-line
  //       this.trackCartItems({ cartItems: items, cartId: cart._id, action: CART_VIEWED }); // eslint-disable-line camelcase
  //     }
  //   }
  //   if (isWidthUp("md", width)) {
  //     // Open the cart, and close after a 3 second delay
  //     openCartWithTimeout(3000);
  //   }
  // };

  /**
   * @name handleSelectOption
   * @summary Called when an option is selected in the option list
   * @private
   * @ignore
   * @param {Object} option The option object that was selected
   * @returns {undefined} No return
   */
  handleSelectOption = (option) => {
    const { sale, uiStore } = this.props;

    // If we are clicking an option, it must be for the current selected variant
    const variant = sale.variants.find((vnt) => vnt._id === uiStore.pdpSelectedVariantId);

    this.selectVariant(variant, option._id);
  };

  // /**
  //  * @name determineSalePrice
  //  * @description Determines a sale's price given the shop's currency code. It will
  //  * use the selected option if available, otherwise it will use the selected variant.
  //  * @returns {Object} An pricing object
  //  */
  // determineSalePrice() {
  //   const { currencyCode, sale } = this.props;
  //   const { pdpSelectedVariantId, pdpSelectedOptionId } = this.props.uiStore;
  //   const selectedVariant = variantById(sale.variants, pdpSelectedVariantId);
  //   let salePrice = {};

  //   if (pdpSelectedOptionId && selectedVariant) {
  //     const selectedOption = variantById(selectedVariant.options, pdpSelectedOptionId);
  //     salePrice = priceByCurrencyCode(currencyCode, selectedOption.pricing);
  //   } else if (!pdpSelectedOptionId && selectedVariant) {
  //     salePrice = priceByCurrencyCode(currencyCode, selectedVariant.pricing);
  //   }

  //   return salePrice;
  // }
  renderTiming() {
    const { sale } = this.props;

    if (sale.hasNotBegun) {
      return (
        <InlineAlert
          alertType="information"
          message={`Begins ${moment(sale.beginsAt).fromNow()}`}
          title="This Can Release has not started"
        />
      );
    }

    if (sale.hasEnded) {
      return (
        <InlineAlert
          alertType="error"
          message={`Ended ${moment(sale.endsAt).fromNow()}`}
          title="This Can Release has ended"
        />
      );
    }

    if (sale.isLowQuantity) {
      return (
        <InlineAlert
          alertType="warning"
          message="Some products have limited availability"
          title="This Can Release is on fire!"
        />
      );
    }

    if (sale.isBackorder) {
      return (
        <InlineAlert
          alertType="error"
          message="Subscribe now to be notified of the next release"
          title="Backorder"
        />
      );
    }

    if (sale.isSoldOut) {
      return (
        <InlineAlert
          alertType="error"
          message="Subscribe now to be notified of the next release"
          title="Sold Out"
        />
      );
    }

    return (
      <InlineAlert
        alertType="warning"
        message={`Sale ends ${moment(sale.endsAt).fromNow()}`}
      />
    );
  }

  renderEmptyGrid() {
    const { classes } = this.props;

    return (
      <Typography className={classes.notFoundMessage} paragraph>
        No products found for this sale.
      </Typography>
    );
  }

  renderProducts() {
    const { sale: { products = [] }, initialSize } = this.props;

    if (products.length === 0) return this.renderEmptyGrid();

    return (
      <Fragment>
        <Grid container spacing={24}>
          <SaleProductGrid
            components={{ SaleProductGridItem }}
            initialSize={initialSize}
            onItemClick={this.onItemClick}
            products={products}
            placeholderImageURL="/static/images/placeholder.gif"
          />
          {/* ...this.props */}
        </Grid>
      </Fragment>
    );
  }

  renderDemoWarning() {
    const {
      sale: {
        isDemo,
        slug: saleSlug,
        shop: { name: shopName, slug: shopSlug } = {}
      } = {}
    } = this.props;

    if (!isDemo) { return; }

    const href =
      `https://www.brewline.io/contact/?utm_source=reserve&utm_medium=shop_sale&utm_campaign=brewer_onboarding&utm_term=${shopSlug}&utm_content=${saleSlug}`;
    const message = (
      <span>
        {`If you represent ${shopName} and would like to run your next can release online, please `}
        <a href={href} target="_blank">contact us</a>!
      </span>
    );

    return (
      <Grid container spacing={24}>
        <Grid item xs={12}>
          <InlineAlert
            alertType="error"
            message={message}
            title="This is a demo can release"
          />
        </Grid>
      </Grid>
    );
  }

  render() {
    const {
      classes,
      sale,
      theme,
      width
    } = this.props;
    const pageTitle = null; // placeholder

    // // Set the default media as the top-level sale's media
    // // (all media on all variants and objects)
    // let pdpMediaItems = sale.media;

    // // If we have a selected variant (we always should)
    // // check to see if media is available, and use this media instead
    // // Revert to original media if variant doesn't have specific media
    // const selectedVariant = sale.variants.find((variant) => variant._id === pdpSelectedVariantId);
    // if (selectedVariant) {
    //   if (selectedVariant.media && selectedVariant.media.length) {
    //     pdpMediaItems = selectedVariant.media;
    //   }

    //   // If we have a selected option, do the same check
    //   // Will revert to variant check if no option media is available
    //   if (Array.isArray(selectedVariant.options) && selectedVariant.options.length) {
    //     const selectedOption = selectedVariant.options.find((option) => option._id === pdpSelectedOptionId);
    //     if (selectedOption) {
    //       if (selectedOption.media && selectedOption.media.length) {
    //         pdpMediaItems = selectedOption.media;
    //       }
    //     }
    //   }
    // }

    // const salePrice = this.determineSalePrice();
    // const compareAtDisplayPrice = (salePrice.compareAtPrice && salePrice.compareAtPrice.displayAmount) || null;

    // Phone size
    if (isWidthDown("sm", width)) {
      return (
        <Fragment>
          {this.renderDemoWarning()}

          <div className={classes.section}>
            <ProductDetailTitle pageTitle={pageTitle} title={sale.headline} />

            <div className={classes.info}>
              <ProductDetailVendor>{sale.shop.name}</ProductDetailVendor>
            </div>

            {this.renderTiming()}
            {/*
            <div className={classes.info}>
              <SaleDetailPrice compareAtPrice={compareAtDisplayPrice} isCompact price={salePrice.displayPrice} />
            </div>
            */}
          </div>

          {/*
          <div className={classes.section}>
            <MediaGallery mediaItems={pdpMediaItems} />
          </div>
          */}

          {this.renderProducts()}

          <div className={classes.section}>
            <ProductDetailDescription>{sale.description}</ProductDetailDescription>
          </div>
        </Fragment>
      );
    }

    return (
      <Fragment>
        {this.renderDemoWarning()}

        <Grid container spacing={theme.spacing.unit * 5}>
          {/* <Grid item className={classes.breadcrumbGrid} xs={12}>
            <Breadcrumbs isPDP tagId={routingStore.tagId} sale={sale} />
          </Grid> */}
          {/* <Grid item xs={12} sm={6}>
            <div className={classes.section}>
              <MediaGallery mediaItems={pdpMediaItems} />
            </div>
          </Grid> */}

          <Grid item xs={12} sm={6}>
            <ProductDetailTitle pageTitle={pageTitle} title={sale.headline} />

            <div className={classes.info}>
              <ProductDetailVendor>{sale.shop.name}</ProductDetailVendor>
            </div>

            {this.renderTiming()}

            {/* <div className={classes.info}>
              <SaleDetailPrice className={classes.bottomMargin} compareAtPrice={compareAtDisplayPrice} price={salePrice.displayPrice} />
            </div> */}

            <div className={classes.info}>
              <ProductDetailDescription>{sale.description}</ProductDetailDescription>
            </div>
          </Grid>

          <Grid item xs={12} sm={6}>
            {this.renderProducts()}
          </Grid>
        </Grid>
      </Fragment>
    );
  }
}

export default SaleDetail;
