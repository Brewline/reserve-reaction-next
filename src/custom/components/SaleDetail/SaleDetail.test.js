// this test file is almost definitely not right. TODO: fix!

import React from "react";
import renderer from "react-test-renderer";
import { MuiThemeProvider } from "@material-ui/core/styles";
import { Provider } from "mobx-react";
import theme from "custom/reactionTheme";
import SaleDetail from "./SaleDetail";
import sampleData from "./__mocks__/saleData.mock";

jest.mock("next/router");

const tags = [
  {
    _id: "v2hE7fzL6cJniwgSm",
    createdAt: "2018-06-11T18:00:48.414Z",
    groups: [],
    isDeleted: false,
    isTopLevel: true,
    isVisible: true,
    name: "Tag A",
    shopId: "J8Bhq3uTtdgwZx3rz",
    slug: "tag-a",
    subTagIds: [
      "s3nybDz4QPkmapETc",
      "JB4FRBiWduNsxhhhn",
      "gRM3ADcY77eNJ7Brm"
    ],
    updatedAt: "2018-06-11T18:01:09.568Z"
  }
];

const routingStore = {
  tagId: "v2hE7fzL6cJniwgSm"
};

const shop = {
  name: "Reaction"
};

const uiStore = {
  pdpSelectedVariantId: "cmVhY3Rpb24vY2F0YWxvZ1Byb2R1Y3RWYXJpYW50OjZxaXFQd0JrZUpkdGRRYzRH",
  setPDPSelectedVariantId: jest.fn().mockName("uiStore.setPDPSelectedVariantId")
};

test("basic snapshot", () => {
  const component = renderer.create((
    <MuiThemeProvider theme={theme}>
      <Provider primaryShopId="J8Bhq3uTtdgwZx3rz" routingStore={routingStore} tags={tags} uiStore={uiStore}>
        <SaleDetail sale={sampleData} shop={shop} currencyCode={"USD"} />
      </Provider>
    </MuiThemeProvider>
  ));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
