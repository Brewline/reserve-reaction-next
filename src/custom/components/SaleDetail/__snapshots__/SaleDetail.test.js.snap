// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
.c0 {
  -webkit-font-smoothing: antialiased;
  color: #3c3c3c;
  font-family: 'Source Sans Pro','Helvetica Neue',Helvetica,sans-serif;
  font-size: 16px;
  font-style: normal;
  font-stretch: normal;
  font-weight: 400;
  -webkit-letter-spacing: .03em;
  -moz-letter-spacing: .03em;
  -ms-letter-spacing: .03em;
  letter-spacing: .03em;
  line-height: 1.5;
  border-radius: 2px;
  border-style: solid;
  border-width: 2px;
  padding-bottom: 15px;
  padding-left: 15px;
  padding-right: 15px;
  padding-top: 15px;
  position: relative;
  white-space: pre-wrap;
  overflow: hidden;
  max-height: 3000vh;
  opacity: 1;
  -webkit-transition: border 0s cubic-bezier(0.785,0.135,0.15,0.86) 250ms,max-height 0s cubic-bezier(0.785,0.135,0.15,0.86) 250ms,padding 0s cubic-bezier(0.785,0.135,0.15,0.86) 250ms,opacity 250ms cubic-bezier(0.785,0.135,0.15,0.86);
  transition: border 0s cubic-bezier(0.785,0.135,0.15,0.86) 250ms,max-height 0s cubic-bezier(0.785,0.135,0.15,0.86) 250ms,padding 0s cubic-bezier(0.785,0.135,0.15,0.86) 250ms,opacity 250ms cubic-bezier(0.785,0.135,0.15,0.86);
  color: #7a6322;
  background-color: #FEF9ED;
  border-color: #E6DBBE;
}

<div
  className="MuiGrid-container-5 MuiGrid-spacing-xs-40-31"
>
  <div
    className="MuiGrid-item-6 MuiGrid-grid-xs-12-45 MuiGrid-grid-sm-6-53"
  >
    <div
      className="SkSaleDetail-info-3"
    />
    <div
      className="c0"
    >
      Sale ends in a few seconds ago
    </div>
    <div
      className="SkSaleDetail-info-3"
    >
      <div
        className="MuiTypography-root-102 MuiTypography-body1-111"
      >
        Sign in as administrator to edit.
You can clone this sale from the sale grid.
You can upload images click or drag in image box on the left here.
Tag this sale below, and then add tag in navigation.
Click the bookmark in the tag to set sale url.
Option variants, price, quantity, and child variants are created by clicking on the variant below, clone the variant to add more options.
Details can be added below the image for more specific sale information.
 Login next to the cart, and then click the dashboard icon for more tools.
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-item-6 MuiGrid-grid-xs-12-45 MuiGrid-grid-sm-6-53"
  >
    <p
      className="MuiTypography-root-102 MuiTypography-body1-111 MuiTypography-paragraph-121 SkSaleDetail-notFoundMessage-4"
    >
      No products found for this sale.
    </p>
  </div>
</div>
`;
